language/compiler/bytecode_source_map/src/source_map.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use failure::prelude::*;
+use ir_to_bytecode_syntax::ast::{Loc, QualifiedModuleIdent, Var_};
+use libra_types::account_address::AccountAddress;
+use libra_types::identifier::Identifier;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::ops::Bound;
+use vm::file_format::{
+    CodeOffset, FieldDefinitionIndex, FunctionDefinitionIndex, StructDefinitionIndex, TableIndex,
+};
+use vm::internals::ModuleIndex;
+
+//***************************************************************************
+// Source location mapping
+//***************************************************************************
+
+pub type SourceMap = Vec<ModuleSourceMap>;
+pub type SourceName = (Identifier, Loc);
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct StructSourceMap {
+    /// The source declaration location of the struct
+    pub decl_location: Loc,
```

Comment:
Does Rust have functors? :p Otherwise can we parameterize this by the type of Loc? 

I can make a separate PR for that if you want, but I'll need it for the source language source information (as its spread across multiple files so doesn't line up with the IR)